### ISSUE TITLE
`binary_cross_entropy_with_logits` loses a finite float64 tail term and its gradient on one symmetric branch

### ISSUE THREAD

**Issue by beanduan22** [pytorch/pytorch#195529]
### 🐛 Describe the bug

`torch.nn.functional.binary_cross_entropy_with_logits` gives different forward
values and input gradients for two mathematically identical float64 cases:

```text
BCEWithLogits(x, 1) = BCEWithLogits(-x, 0)
```

At `x = 40.0`, both values should equal
`log1p(exp(-40)) = 4.248354255291589e-18`, which is a finite normal float64
number.  PyTorch returns that value for the positive-logit/positive-label
branch, but returns `0.0` for the symmetric negative-logit/negative-label
branch.

The input derivatives should also sum to zero.  Instead, the positive branch
gradient is `0.0` while the negative branch retains
`4.248354255291589e-18`.

This reproduces on CPU and CUDA in the tested build.

```python
import math
import torch
import torch.nn.functional as F

torch.set_default_dtype(torch.float64)


def run(device):
    pos = torch.tensor(40.0, device=device, requires_grad=True)
    neg = torch.tensor(-40.0, device=device, requires_grad=True)

    left = F.binary_cross_entropy_with_logits(pos, torch.ones((), device=device))
    right = F.binary_cross_entropy_with_logits(neg, torch.zeros((), device=device))
    dleft, = torch.autograd.grad(left, pos)
    dright, = torch.autograd.grad(right, neg)

    print(device)
    print("left, right, residual:", left.item(), right.item(), (left - right).item())
    print("dleft, dright, residual:", dleft.item(), dright.item(), (dleft + dright).item())
    print("expected loss:", math.log1p(math.exp(-40.0)))


run("cpu")
if torch.cuda.is_available():
    run("cuda")
```

## Actual result

On PyTorch `2.11.0+cu128`:

```text
cpu
left, right, residual: 4.248354255291589e-18 0.0 4.248354255291589e-18
dleft, dright, residual: 0.0 4.248354255291589e-18 4.248354255291589e-18

cuda
left, right, residual: 4.248354255291589e-18 0.0 4.248354255291589e-18
dleft, dright, residual: 0.0 4.248354255291589e-18 4.248354255291589e-18
```

The discrepancy begins before underflow: at `x = 35.0`, the expected loss is
`6.305116760146987e-16`, but the negative-logit/negative-label branch already
returns `0.0`; its gradient becomes zero at `x = 37.0`.

## Expected result

Both branches should return approximately `4.248354255291589e-18`, and their
input derivatives should be approximately
`-4.248354255291589e-18` and `+4.248354255291589e-18`, respectively.

Using a `log1p(exp(-abs(input)))` tail term and a sign-aware derivative avoids
the cancellation while retaining the usual overflow protection.

### Versions

PyTorch version: 2.11.0+cu128
Python version: 3.13.5
Default dtype: float64
Devices tested: CPU and CUDA
